ws/DiscordWebsocket.ts

```diff
@@ -0,0 +1,229 @@
+    this.connectionProperties = { $os: process.platform, $browser: 'dis.ts', $device: 'dis.ts', ...connProps };
+  }
+
+  connect() {
+    this.initialize();
+    return this;
+  }
+
+  disconnect(code?: number) {
+    if (code === 1000 || code === 1001) {
+      this.sessionID = null;
+      this._seq = null;
+    }
+    this._selfDisconnect = true;
+    this.ws?.close(code);
+    return this;
+  }
+
+  initialize() {
+    if (!this._url) throw new Error('Websocket URL not provided');
+
+    this._selfDisconnect = false;
+    if (this.ws) throw new Error('Attempted duplicate connection - run uninitialize');
+
+    this.ws = new WebSocket(this._url);
+    this.ws.once('close', this._onClose);
+    this.ws.on('message', this._onMessage);
+    this.ws.on('error', console.error);
+    return this;
+  }
+
+  reset() {
+    if (this._heartBeatInterval) clearInterval(this._heartBeatInterval);
+    if (this.ws?.readyState === WebSocket.OPEN) {
+      this.ws.off('message', this._onMessage)
+        .off('error', console.error)
+        .terminate();
+    }
+    this._heartBeatInterval = null;
+    this.ws = null;
+    this.ready = false;
+    this._lastHeartbeatAck = true;
+    return this;
+  }
+
+  restart(newSession = false) {
+    if (newSession) this.sessionID = null;
+    this.disconnect().connect();
+  }
+
+  // TODO Prevent 4012, 4013?
+  send(
+    op: GATEWAY_OPCODES.IDENTIFY | GATEWAY_OPCODES.RESUME | GATEWAY_OPCODES.HEARTBEAT | GATEWAY_OPCODES.REQUEST_GUILD_MEMBERS | GATEWAY_OPCODES.VOICE_STATE_UPDATE | GATEWAY_OPCODES.PRESENCE_UPDATE,
+    d: any,
+    override = false,
+  ) {
+    if (this.ws?.readyState !== WebSocket.OPEN) return false;
+    if (override
+      || (this.ws?.readyState === WebSocket.OPEN
+        && ((this.ready && !(op === GATEWAY_OPCODES.IDENTIFY || op === GATEWAY_OPCODES.RESUME))
+          || (!this.ready && (op === GATEWAY_OPCODES.IDENTIFY || op === GATEWAY_OPCODES.RESUME))
+        )
+      )
+    ) {
+      this.ws.send(JSON.stringify({ op, d }));
+      return true;
+    }
+    return false;
+  }
+
+  setIntents(intents: number) {
+    this._intents = intents;
+  }
+
+  setToken(token: string) {
+    this._token = token;
+  }
+
+  setURL(url: string) {
+    this._url = url;
+  }
+
+  private _heartbeat(scheduled = false) {
+    if (scheduled) {
+      if (!this._lastHeartbeatAck) {
+        console.error(new Error('No heartbeat acknowledgement'));
+        return this.restart(true);
+      }
+      this._lastHeartbeatAck = false;
+    }
+    this.send(GATEWAY_OPCODES.HEARTBEAT, this._seq, true);
+  }
+
+  private _hello(d: { _trace: string; heartbeat_interval: number }) {
+    this._heartBeatInterval = setInterval(() => this._heartbeat(), d.heartbeat_interval);
+    this.discordTrace = d._trace;
+
+    if (this.sessionID) this._resume();
+    else this._identify();
+  }
+
+  // TODO max_concurrency implemetation if necessary
+  // TODO prevent identifying without a token
```

Comment:
if there are no token, same thing => erroring out.

For max_concurrency, we shoudl left it out until we have something entirely working. Same for compression.
That can be added later in separate PRs.